### PR TITLE
[improve][function] Get function class name from the NAR when using built-in functions

### PR DESCRIPTION
**Pull request by cbornet**
### Motivation

Currently the `--classname` param is required when creating a function and the functionClass value of the `pulsar-io.yaml` of the NAR is not used.
This change is about reading `functionClass` from the NAR and using it if the classname is not provided the same way we have for connectors.

### Modifications

* Allow to not provide the classname in the admin CLI if the Function is built-in.
* Replace `Functions` by `TreeMap<String, FunctionArchive>` where `FunctionArchive` contains the archive path, the nar class loader and the `FunctionDefinition` in `FunctionsManager` and in `searchForFunctions`'s return.
* Close the class loader in `validateUpdateRequestParams` when not using builtin functions
* Add the possibility to configure the local runner functions directory

### Verifying this change

- [x] Make sure that the change passes the CI checks.

This change added tests and can be verified as follows:

* Run `PulsarFunctionE2ETest::testPulsarFunctionBuiltin`
* Run `PulsarFunctionE2ETest::testPulsarFunctionBuiltinDoesNotExist`
* Run `PulsarFunctionLocalRunTest:testE2EPulsarFunctionLocalRunBuiltin`

### Does this pull request potentially affect one of the following parts:

No

*If `yes` was chosen, please highlight the changes*

  - Dependencies (does it add or upgrade a dependency): (yes / no)
  - The public API: (yes / no)
  - The schema: (yes / no / don't know)
  - The default values of configurations: (yes / no)
  - The wire protocol: (yes / no)
  - The rest endpoints: (yes / no)
  - The admin cli options: (yes / no)
  - Anything that affects deployment: (yes / no / don't know)

### Documentation

Check the box below or label this PR directly.

Need to update docs? 

- [ ] `doc-required` 
(Your PR needs to update docs and you will update later)
  
- [x] `no-need-doc` 
doc on built-in functions doesn't exist at all
  
- [ ] `doc` 
(Your PR contains doc changes)

- [ ] `doc-added`
(Docs have been already added)

### CODE REVIEW

**Comment by cbornet** [comment]
Added tests so this PR is now ready to be reviewed.

**Review by eolivelli — APPROVED** [review]
LGTM

great work

File: pulsar-client-tools/src/main/java/org/apache/pulsar/admin/cli/CmdFunctions.java [hunk 1]
```diff
@@ -643,14 +643,18 @@ void processArguments() throws Exception {
 
         protected void validateFunctionConfigs(FunctionConfig functionConfig) {
             // go doesn't need className
-            if (functionConfig.getPy() != null || functionConfig.getJar() != null) {
+            if (functionConfig.getPy() != null
+                    || (functionConfig.getJar() != null && !functionConfig.getJar().startsWith("builtin://"))) {
```
Comment (shibd) [inline pulsar-client-tools/src/main/java/org/apache/pulsar/admin/cli/CmdFunctions.java:647]: Add unit test coverd it.

File: pulsar-functions/utils/src/main/java/org/apache/pulsar/functions/utils/functions/FunctionUtils.java [hunk 2]
```diff
@@ -98,24 +95,29 @@ public static Functions searchForFunctions(String functionsDirectory, boolean al
...
-                    log.error(cntDef.getFunctionClass());
+
+                    functionArchiveBuilder.archivePath(archive);
+
+                    functionArchiveBuilder.classLoader(ncl);
+                    functionArchiveBuilder.functionDefinition(cntDef);
+
                     if (alwaysPopulatePath || !StringUtils.isEmpty(cntDef.getFunctionClass())) {
-                        functions.functions.put(cntDef.getName(), archive);
+                        functions.put(cntDef.getName(), functionArchiveBuilder.build());
```
Comment (shibd) [inline pulsar-functions/utils/src/main/java/org/apache/pulsar/functions/utils/functions/FunctionUtils.java:113]: When `alwayPopulatePath` == false or `cntDef.getFunctionClass()` is empty, function may be result empty. This is not the same as the original logic.  Please check if there is a problem, and add a unit test to cover it.

**Review by shibd — COMMENTED** [review]
Let some comments.

File: pulsar-functions/utils/src/main/java/org/apache/pulsar/functions/utils/functions/FunctionUtils.java (hunk 2 above)
Comment (cbornet) [reply pulsar-functions/utils/src/main/java/org/apache/pulsar/functions/utils/functions/FunctionUtils.java:113]: The logic didn't change here (the condition is untouched). And there are already tests covering this. Or am I missing something ?

File: pulsar-functions/utils/src/main/java/org/apache/pulsar/functions/utils/functions/FunctionUtils.java [hunk 3]
```diff
@@ -98,24 +95,29 @@ public static Functions searchForFunctions(String functionsDirectory, boolean al
...
                     }
-
-                    functions.functionsDefinitions.add(cntDef);
```
Comment (shibd) [inline pulsar-functions/utils/src/main/java/org/apache/pulsar/functions/utils/functions/FunctionUtils.java:109]: > The logic didn't change here (the condition is untouched).

You removed this line.

File: pulsar-functions/utils/src/main/java/org/apache/pulsar/functions/utils/functions/FunctionUtils.java (hunk 2 above)
Comment (shibd) [reply pulsar-functions/utils/src/main/java/org/apache/pulsar/functions/utils/functions/FunctionUtils.java:113]: https://github.com/apache/pulsar/pull/15693/files/2e9738aacb24e56571a0c7f9a629841b6290d532#r886692769

**Review by shibd — APPROVED** [review]
/LGTM

File: pulsar-functions/utils/src/main/java/org/apache/pulsar/functions/utils/functions/FunctionUtils.java [hunk 4]
```diff
@@ -98,24 +95,29 @@ public static Functions searchForFunctions(String functionsDirectory, boolean al
         try (DirectoryStream<Path> stream = Files.newDirectoryStream(path, "*.nar")) {
             for (Path archive : stream) {
                 try {
-                    FunctionDefinition cntDef = FunctionUtils.getFunctionDefinition(archive.toString());
+
+                    NarClassLoader ncl = NarClassLoaderBuilder.builder()
```
[the remark this answers is not in the document]
Comment (cbornet) [reply pulsar-functions/utils/src/main/java/org/apache/pulsar/functions/utils/functions/FunctionUtils.java:99]: This code is inspired by what is done for [Sources/Sinks](https://github.com/apache/pulsar/blob/9c5bc38144ea3fb74564d9291f3073ab94ebdf41/pulsar-functions/utils/src/main/java/org/apache/pulsar/functions/utils/io/ConnectorUtils.java#L155). For connectors, the NAR class loaders are also stored in the ConnectorsManager and never closed.

**Review by nicoloboschi — APPROVED** [review]
LGTM

**Pull request merged**